### PR TITLE
diff: add a file-by-file variant for external diff tools

### PR DESCRIPTION
**Pull request by fowles**
# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

**Comment by yuja** [comment]
Just for reference, there are some discussion in https://github.com/martinvonz/jj/issues/3696
(I haven't reviewed the implementation yet)

File: cli/src/diff_util.rs
Comment (yuja) [inline cli/src/diff_util.rs:91]: Perhaps, `--tool-mode` shouldn't require `--tool`, and it has to be `Option<DiffToolMode>`, because:
* it makes sense to set the default tool in user config, and select mode at runtime.
* if the tool doesn't support dir-diff, the default would fall back to file-by-file.

File: cli/src/diff_util.rs
Comment (yuja) [inline cli/src/diff_util.rs:738]: Can't we reuse `check_out_trees()` to materialize files all at once, and execute diff tool for each file pair? That will help extend file-by-file mode to something like `tool -- file1a file1b -- file2a file2b` or `tool file1a file1b &; tool file2a file2b &;`.
https://github.com/martinvonz/jj/issues/3696#issuecomment-2116442540

File: cli/src/config-schema.json
Comment (yuja) [inline cli/src/config-schema.json:122]: I think we'll also need per-tool default/supported modes or per-tool/mode args.
https://github.com/martinvonz/jj/issues/3696#issuecomment-2116426056

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs:738]: I want to avoid a full checkout because the way I plan to do file copy tracking so that it works for all our tools is in the materialize method we are using.

Comment (yuja) [reply cli/src/diff_util.rs:738]: Can't we port `check_out_trees()` to be based on the diff materialization logic then? It'll be nice if `diffedit --tool` can somehow deal with file renames.

Comment (fowles) [reply cli/src/diff_util.rs:738]: I don't think so because the trees need to have files with actual names in them that makes such a mapping difficult.  How would you represent a copy cycle of A -> B -> C ->A for example.

Comment (yuja) [reply cli/src/diff_util.rs:738]: Maybe I don't understand the underlying problem, but `check_out_trees()` first calculates `left_tree.diff_stream(right_tree, matcher)` to build a sparse pattern, then checks out files from `left`/`right_trees`. I assumed it's roughly equivalent to dump file pairs yielded by `materialized_diff_stream()`.

Comment (fowles) [reply cli/src/diff_util.rs:738]: when you create the directory structure it most represent both sides for all file simulatenously.  The stream only has to represent both sides for each file one at a time.  As a result, a stream can handle oddities like cycles of renames where the directory cannot.

File: cli/src/config-schema.json
Comment (fowles) [reply cli/src/config-schema.json]: Good idea, switch to per-tool only.

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs]: I made `--tool-mode` optional and it will override the configured value for a tool, but it is still required to name the tool, because otherwise the usage gets confusing.

File: docs/config.md
Comment (ilyagr) [inline docs/config.md:206]: I haven't thoroughly checked, but is this example out-of-date now that the config is per-tool?

File: cli/src/diff_util.rs
Comment (ilyagr) [inline cli/src/diff_util.rs:91]: Optional: I'm a bit torn on whether this option is a good idea, but if we do it, how about:

```
jj diff --file-by-file-tool pratdiff
```

(or `--tool-file-by-file`; I also considered `--file-by-file --tool pratdiff`)

Note that this doesn't actually work very well: with this setup, `pratdiff` does not show color, and I couldn't find a way to fix it without creating a diff config for `pratdiff`. So, maybe just having a config know is sufficient.

File: cli/src/diff_util.rs
Comment (ilyagr) [inline cli/src/diff_util.rs:714]: I would print some sort of heading here. I think that *at least* a newline is called for. But perhaps we should do something nicer:

> (a newline)
> JJ: Comparing XXX and YYY

Either JJ:, or the whole line, or the whole line except for the paths could be in yellow.

More questionably, in the common case you could print:

> Comparing {left,right}/PATH

Ideally, the heading would become templated at some point.

This is also important if the diff is opened in a separate window and `jj` waits for the program to finish. (Guess on Windows, people will just get 15 windows at once unless they take special care).

Comment (ilyagr) [reply cli/src/diff_util.rs:714]: Also, adding `#5 of 20` would be helpful.

File: docs/config.md
Comment (ilyagr) [reply docs/config.md:206]: OK, it's supposed to be

```
[merge-tools.pratdiff]
diff-args = ["--color=always", "$left", "$right"]
mode = "file-by-file"
```

but it's not a good example since the mode doesn't change anything.

Perhaps vimdiff would be a better example. In fact, we could use the file-by-file mode by default there.

File: cli/src/config-schema.json
Comment (yuja) [inline cli/src/config-schema.json:352]: nit: should be renamed to `diff-mode` or something because it only applies to `diff-args`.

btw, do you have any opinion about `file-diff-args`/`dir-diff-args` stuff?
https://github.com/martinvonz/jj/issues/3696#issuecomment-2116426056

(Even if we add `file`/`dir-diff-args`, it might make sense to add mode flag or config option. There might be `file-by-file` and `files-all-at-once` modes that share the same `file-diff-args`.)

File: cli/src/diff_util.rs
Comment (yuja) [reply cli/src/diff_util.rs]: Well, we have a similar problem with `--context` option. A better way to handle these is to warn/error if the selected tool doesn't support these command-line options.

File: cli/src/diff_util.rs
Comment (yuja) [reply cli/src/diff_util.rs:724]: I don't quite understand the reasoning, sorry. Perhaps, this part should be reviewed by someone else who have expertise on copy tracing.

> The stream only has to represent both sides for each file one at a time. As a result, a stream can handle oddities like cycles of renames where the directory cannot.

Does it mean we can't support `files-all-at-once` mode (in which jj will materialize all changed file pairs and spawn diff tool with file name pairs e.g. `tool a/foo b/foo &; tool a/bar b/renamed-bar &; ...`)?

Comment (ilyagr) [reply cli/src/diff_util.rs:724]: I also am a little confused about why we couldn't just make two directories with all the correct files (but I'm also not an expert on copy tracking). If a rename cycle is so complicated that the "before" and "after" states cannot be represented correctly, perhaps we should prohibit it entirely?

Do you have an example where it's a problem?

As an aside, I don't think this PR is a prerequisite for copy tracking, or at least I'm missing the reason why it would be. At first, it would be OK if only the internal diff knew how to pair the files.

File: cli/src/config-schema.json
Comment (ilyagr) [reply cli/src/config-schema.json:352]: +1 on including `diff` in the setting name.

I would even go further: I think we should avoid non-descript `mode` if we can. Who knows what other settings that could be called "mode" will exist in the future.

Perhaps `diff-dir-mode`? (to be joined by `edit-dir-mode`, etc, in the future). I'd prefer to come up with something even better, but this is what I have now.

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs:724]: At a high level, the directory approach doesn't have a good mechanism for copy tracking because there is no way to communicate to the external tool what the mapping of files is.  So you have two options:

1. lay out the directories as they actually exist at each version and then be unable to tell the tool what the copies are (status quo)
2. lay out the directories using a set of names that moves around old files based on where the new files actually end up so that external tool will show the right thing (won't properly handle copy cycles and also feels error prone on the logic side) 

If you have a thing where to files swap names (easier to think about then the three way cycle).  how do you suggest laying out the directory structure?  I am fairly certain there is not a simple answer that will make it work.

The PR is not a hard prereq to using copy tracking in any diff stuff, but it is a pre-req to external diff tools support for copy tracking given the above.  Since my explicit interest in this is with an external diff tool, I don't feel bad about pushing it because of that.

Comment (martinvonz) [reply cli/src/diff_util.rs:724]: > If you have a thing where to files swap names (easier to think about then the three way cycle). how do you suggest laying out the directory structure?

Let's X was renamed to Y and Y was renamed to X.Let's say that happened in commit B, which is on top of commit A. Then I would expect a diff files X and Y, where the LHS of X is Y@A and the RHS isX@B, and where the LHS of Y is X@A and the the RHS is Y@B. Should be possible, right?

It can still be confusing to the user. To help with that, we can print a message to the terminal, or maybe to a temporary file (like the `JJ-INSTRUCTIONS` file we already have).

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs]: Removed the argument and made this a config only setting.

File: docs/config.md
Comment (fowles) [reply docs/config.md:206]: Switched to vimdiff and renamed mode to `diff-mode`.

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs:91]: I removed the argument and made this a config only setting

File: cli/src/config-schema.json
Comment (fowles) [reply cli/src/config-schema.json:352]: Done

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs:684]: Added a helper in file_util

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs:720]: the only way to accomplish what you are suggesting in the directory structure is to have the thing in LEFT not be an actual checkout but to have it be a checkout with some renames applied, which is weird IMO.

File-by-file on the other hand this is a very natural representation.

Comment (ilyagr) [reply cli/src/diff_util.rs:720]: I was thinking of `file-by-file` making the same checkout as normal, and then calling the tool file-by-file. Future methods could provide the pairings to the tool in other ways (pairs of arguments, text file).

Even for dirdiff, you can describe the correspondence in JJ-INSTRUCTIONS, as I think Martin suggested just above.


> lay out the directories as they actually exist at each version and then be unable to tell the tool what the copies are (status quo)

For the minimal dirs, there is some advantage move/copy tracking can provide you. If file A was copied to B, you can put A on the LHS and B on the RHS (as opposed to putting nothing on the LHS as would happen now).

Comment (fowles) [reply cli/src/diff_util.rs:720]: that will end up replicating a lot of the logic for copy detection in a way that feel awkward to me.  I think the proper place to isolate that logic is inside materialize_diff_stream which will allow for common handling across all our diff types

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs]: I added this, but I am not sure that I actually like it.  I will go with whatever the crowd decides here though

File: cli/src/diff_util.rs
Comment (ilyagr) [inline cli/src/diff_util.rs:722]: Could you put another newline before `JJ`?

--------------

My reasoning for the header is to make it clearer what's going on

- if there's an error
- if a GUI tool is called each time

It's not perfect, and perhaps other people will have other opinions.

Comment (fowles) [reply cli/src/diff_util.rs]: Done

File: cli/src/config-schema.json
Comment (ilyagr) [inline cli/src/config-schema.json:335]: Do people have an opinion on `diff-mode` vs `diff-dir-mode`? I prefer the latter, in case there are other kind of options that could be named "mode" are added in the future, but maybe people have better ideas or are fine with `diff-mode`?

We can also rename it in a later PR (especially if this is merged after the release).

File: cli/src/diff_util.rs
Comment (ilyagr) [reply cli/src/diff_util.rs]: Are you trying to get this PR into the Wednesday release?

In that case, if we can't get to the same page, one possibility would be to add `experimental-` to the option name, expecting future breaking changes (if Martin/Yuya agree).

As far as I'm concerned, I don't want to block your progress, but I'm not sure there is a consensus here. I'm, personally, not entirely opposed to creating files one-by-one, but I still don't understand the decision to do that very well (it could be just a matter of time).

If Martin or Yuya are on the same page as you, I'm OK with this.

Comment (fowles) [reply cli/src/diff_util.rs]: no, I am happy to miss the release since the thing I really want is the integration that is a next step

Comment (fowles) [reply cli/src/diff_util.rs]: I have a pretty clear idea in my mind of how I want to use this, so I think it will become more clear in a follow up CL when I integrate these things into the copy detection work

File: cli/src/config-schema.json
Comment (fowles) [reply cli/src/config-schema.json:335]: I don't like `diff-dir-mode` because that is only one of they types.  `diff-dir-more=file-by-file` seems poorly named to me.

Comment (ilyagr) [reply cli/src/config-schema.json:335]: Yeah, in my mind this means "the approach we take to the fact that the diff contains a whole dir is that we send the dirs to the tool one file at a time".

I'd like a better name, but I don't think this is blocking; we can change it later (we'll have a month to do it without a breaking change if we wait until Wednesday to merge).

File: cli/src/diff_util.rs
Comment (ilyagr) [reply cli/src/diff_util.rs]: Oh, I made a silly error. We shouldn't print that newline on the first iteration.

File: cli/src/config-schema.json
Comment (joyously) [reply cli/src/config-schema.json:335]: `diff-dir-singly`? `diff-each-file-in-dir`?

File: cli/src/diff_util.rs
Comment (yuja) [reply cli/src/diff_util.rs]: > I was thinking of `file-by-file` making the same checkout as normal, and then calling the tool file-by-file. Future methods could provide the pairings to the tool in other ways (pairs of arguments, text file).

I was considering this, too. If the diff tool is GUI/CUI-based, user might want to open all pairs at once.

And I assume the implementation can be switched to be based off `materialize_diff_stream()` if that makes sense, something like:
```rust
let mut file_name_pairs: Vec<(RepoPath, RepoPath)>;
while let Some(((left_path, right_path), (left_value, right_value))) = diff_stream.next().await {
    write_file(left_temp_dir, left_path, left_value);
    write_file(right_temp_dir, right_path, right_value);
    file_name_pairs.push((left_path, right_path));
}
if diffedit {
    somehow_construct_tree_state_for_snapshot(right_temp_dir, file_name_pairs); // XXX
}
for (left_path, right_path) in file_name_pairs {
    invoke_external_tool(..);
    if !batch {
        wait();
    }
}
```

File: cli/src/config-schema.json
Comment (yuja) [reply cli/src/config-schema.json:335]: "diff-(tool-)exec-mode" or "diff-(tool-)invocation-mode"? ("tool" is omitted because it's under the "merge-tools" section.)

File: cli/src/diff_util.rs
Comment (yuja) [reply cli/src/diff_util.rs]: If we print header by ourselves, I would copy the one from color-words/git diffs. I'm not sure if the header should be emitted by default. The diff command usually prints some header, so it's redundant if the tool output is streamed.

Comment (ilyagr) [reply cli/src/diff_util.rs]: I elaborated a bit in https://github.com/martinvonz/jj/pull/3982#discussion_r1660300300. If the program just outputs some text and prints some header, I'd agree. If it's more like `code --wait`, I'm not so sure.

Comment (yuja) [reply cli/src/diff_util.rs]: Maybe that's another reason we'll need `type = stream|gui|tui`.

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs]: Done

File: cli/src/config-schema.json
Comment (fowles) [reply cli/src/config-schema.json:335]: went with diff-invocation-mode

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs:710]: stripped out headers for now and will offer a separate PR just for headers

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs:720]: deferring this discussion until later work that integrates copy-tracking.  I think that will make it more clear what the best answer is

File: lib/src/file_util.rs
Comment (yuja) [inline lib/src/file_util.rs:65]: maybe you forgot to extract `create_parent_dirs()` from local_working_copy.rs?

(and it's preferred to do that in separate commit)

File: cli/src/diff_util.rs
Comment (yuja) [inline cli/src/diff_util.rs:666]: nit: is it possible (and make sense) to move this to `merge_tools/external.rs`?

(Maybe we can move `materialized_diff_stream()` to library, but I'm not sure about `diff_content()`.)

File: cli/src/diff_util.rs
Comment (yuja) [inline cli/src/diff_util.rs:683]: nit: `fs::write()` can be used

File: cli/src/merge_tools/external.rs
Comment (yuja) [inline cli/src/merge_tools/external.rs:56]: nit: `clap::ValueEnum` is no longer needed

File: cli/src/merge_tools/external.rs
Comment (yuja) [inline cli/src/merge_tools/external.rs:296]: nit: take `patterns` as ref

File: CHANGELOG.md
Comment (yuja) [inline CHANGELOG.md:82]: (reminder) you'll need to move this entry after Wednesday's release.

**Review by yuja — APPROVED** [review]
Thanks.

File: lib/src/file_util.rs
Comment (fowles) [reply lib/src/file_util.rs:65]: I didn't realize what you were suggesting.  The error types are different so the entire thing is a bit awkward.  I will do it in a separate commit.

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs:683]: done

File: cli/src/merge_tools/external.rs
Comment (fowles) [reply cli/src/merge_tools/external.rs:56]: done

File: cli/src/merge_tools/external.rs
Comment (fowles) [reply cli/src/merge_tools/external.rs:296]: done

File: cli/src/diff_util.rs
Comment (fowles) [reply cli/src/diff_util.rs:666]: I tried moving stuff around a bit, but the changes don't make it much cleaner.  Let me see if I can improve it on the next change in this sequence.

File: lib/src/file_util.rs
Comment (yuja) [reply lib/src/file_util.rs:65]: Indeed. `create_parent_dirs()` will have to return `io::Error` or something.

Can you remove `file_util::create_dir_all()` then? It doesn't make sense to wrap std fn. The function interface will be different anyway (because `create_parent_dirs()` ensures that no directories are created outside of the base directory.)

Comment (fowles) [reply lib/src/file_util.rs:65]: Done

File: CHANGELOG.md
Comment (fowles) [reply CHANGELOG.md:90]: Done

**Pull request merged**
